weave/Scripts/Main.cs

```diff
@@ -1,38 +1,82 @@
+using System.Collections.Generic;
+using System.Linq;
 using Godot;
 using GodotSharper.AutoGetNode;
+using GodotSharper.Instancing;
+using weave.Utils;
+
+namespace weave;
 
 public partial class Main : Node2D
 {
-    [GetNode("MegaCoolLabel")]
-    private Label _label;
-
-    [GetNode("KillArea")]
-    private Area2D _area2D;
+    private readonly ISet<Player> _players = new HashSet<Player>();
+    private int _nPlayers = 3;
 
-    [GetNode("Player")]
-    private Player _player;
+    /// <summary>
+    ///     How many players that have reached the goal during the current round.
+    /// </summary>
+    private int _roundCompletions;
 
     public override void _Ready()
     {
         this.GetNodes();
-        _label.Text = "Hello World!";
-
-        _area2D.BodyEntered += OnAreaEntered;
-        _player.PlayerShotBullet += OnPlayerShotBullet;
+        SpawnPlayers(_nPlayers);
+        ClearAndSpawnGoals();
     }
 
-    private void OnPlayerShotBullet(Node2D bullet, Vector2 globalPosition)
+    private void SpawnPlayers(int amount)
     {
-        AddChild(bullet);
-        bullet.GlobalPosition = globalPosition;
+        amount.TimesDo(() =>
+        {
+            var playerId = UniqueId.Generate();
+            var player = Instanter.Instantiate<Player>();
+            _players.Add(player);
+
+            AddChild(player);
+            player.GlobalPosition = GetRandomCoordinateInView(100);
+            player.PlayerId = playerId;
+        });
     }
 
-    private void OnAreaEntered(Node2D node)
+    private void OnPlayerReachedGoal(Player player)
     {
-        if (node is Player player)
+        GD.Print($"Player {player.PlayerId} has reached the goal");
+        _roundCompletions++;
+
+        if (_roundCompletions == _nPlayers)
         {
-            GD.Print("player entered the area");
-            player.QueueFree();
+            GD.Print("All players have reached the goal");
+
+            _roundCompletions = 0;
+            ClearAndSpawnGoals();
         }
     }
+
+    private void ClearAndSpawnGoals()
+    {
+        // Remove existing goals
+        GetTree()
+            .GetNodesInGroup(GroupConstants.GoalGroup)
+            .ToList()
+            .ForEach(goal => goal.QueueFree());
+
+        // Spawn new goals
+        _players
+            .ToList()
+            .ForEach(player =>
+            {
+                var goal = Instanter.Instantiate<Goal>();
+                CallDeferred("add_child", goal);
+                goal.GlobalPosition = GetRandomCoordinateInView(100);
+                goal.PlayerReachedGoal += OnPlayerReachedGoal;
+                goal.CallDeferred("set", nameof(Player.PlayerId), player.PlayerId);
+            });
+    }
+
+    private Vector2 GetRandomCoordinateInView(float margin)
```

Comment:
In the future this method (or some other method) should not be able to give the same coordinates that another entity has. Sometimes the player spawns on top of their goal

Reply:
Yes, this is just a placeholder